docker/test_docker.py

```diff
@@ -0,0 +1,589 @@
+            # but writable only by the owner
+            (td_path / "packages").mkdir(mode=0o444)
+
+            proc = subprocess.Popen(
+                (
+                    "docker",
+                    "run",
+                    "--rm",
+                    *extra_args,
+                    "-v",
+                    # Mount the temporary directory as the /data directory
+                    f"{tmpdir}:/data",
+                    image,
+                ),
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+            _, err = proc.communicate()
+            # we should error out in this case
+            assert proc.returncode != 0
+            assert "must have read/write/execute access" in err.decode()
+
+    def test_runs_as_pypiserver_user(self, image: str) -> None:
+        """Test that the default run uses the pypiserver user."""
+        proc = subprocess.Popen(
+            (
+                "docker",
+                "run",
+                "--rm",
+                "--detach",
+                "--publish",
+                "8080:8080",
+                image,
+            ),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        out, err = proc.communicate()
+        # we should error out in this case
+        assert proc.returncode == 0, err and f"{err.decode()}"
+        container_id = out.decode().strip()
+        try:
+            wait_for_container(8080)
+            proc = subprocess.Popen(
+                ("docker", "container", "exec", container_id, "ps", "a"),
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+            out, err = proc.communicate()
+            assert proc.returncode == 0, err and f"{err.decode()}"
+            proc_line = next(
+                filter(
+                    # grab the process line for the pypi-server process
+                    lambda ln: "pypi-server" in ln,
+                    map(bytes.decode, out.splitlines()),
+                )
+            )
+            user = proc_line.split()[1]
+            # the ps command on these alpine containers doesn't always show the
+            # full user name, so we only check for the first bit
+            assert user.startswith("pypi")
+        finally:
+            subprocess.Popen(
+                ("docker", "container", "rm", "-f", container_id)
+            ).communicate()
+
+
+class TestBasics:
+    """Test basic pypiserver functionality in a simple unauthed container."""
+
+    @pytest.fixture(scope="class")
+    def container(self, image: str) -> t.Iterator[str]:
+        """Run the pypiserver container.
+
+        Returns the container ID.
+        """
+        proc = subprocess.Popen(
+            (
+                "docker",
+                "run",
+                "--rm",
+                "--publish",
+                "8080:8080",
```

Comment:
Can we not bind on 8080 on the host but on a random port, someone running the tests may have something else running on that port already